### PR TITLE
Handle stream pipeline errors

### PR DESCRIPTION
**Pull request by robertkiel**
## Context:

Stream can internally produce multiple errors such as:
- the socket (the ultimate sink) produces a write error
- the handling of the message lead to an error case
- reading from a socket fails

## Previous behavior

The error led to an unhandled promise rejection and since Node.js 15 (see https://nodejs.medium.com/node-js-v15-0-0-is-here-deb00750f278 ) are treated as errors, the process terminates on such kind of errors.

## New behavior

Add a try / catch block and forward errors such that they can be handled appropriately

## Tasks:

- [x] add try / catch block to client-side relay connection
- [x] add try / catch block to relay-side relay connection
- [x] add try / catch block to webRTC connection

## Other fixes

- always switch connection, even if there is no reconnect handler or reconnect handler produces an error
- don't call TCP done handler twice
- fix debug logs

## Related errors

- hoprnet/hoprnet#2331
- hoprnet/hoprnet#2321
- hoprnet/hoprnet#2314
- hoprnet/hoprnet#2322

### CODE REVIEW

File: src/relay/connection.spec.ts [hunk 1]
```diff
@@ -463,3 +458,87 @@ describe('relay connection', function () {
...
+        })()
+      ),
+      Error(errorInSource)
+    )
+
+    await sourcePromise
+  })
+
+  it('correct sources in falsy sinks', async function () {
+    // const AliceRelay = Pair<StreamType>()
```
Comment (andreykiryushkin) [inline src/relay/connection.spec.ts:500]: please remove

File: src/relay/connection.ts [hunk 2]
```diff
@@ -159,9 +159,35 @@ class RelayConnection extends EventEmitter implements MultiaddrConnection {
...
+      sinkCreator.catch(deferred.reject)
+
+      this._sinkSourceAttached = true
+      this._sinkSourceAttachedPromise.resolve(
+        async function* (this: RelayConnection) {
+          try {
+            yield* toU8aStream(source)
+            deferred.resolve()
+          } catch (err: any) {
+            console.log(`error caught`, err.message)
```
Comment (andreykiryushkin) [inline src/relay/connection.ts:178]: use `this.log` of `RelayConnection`

**Review by andreykiryushkin — CHANGES_REQUESTED** [review]

File: src/webrtc/connection.ts [hunk 3]
```diff
@@ -18,8 +18,8 @@ import type { Stream, StreamResult, DialOptions } from '../types'
 const DEBUG_PREFIX = `hopr-connect`
 
 const _log = Debug(DEBUG_PREFIX)
-const _verbose = Debug('hopr-connect:verbose')
-const _error = Debug(DEBUG_PREFIX.concat(`error`))
+const _verbose = Debug(DEBUG_PREFIX.concat(':verbose'))
```
Comment (tolbrino) [inline src/webrtc/connection.ts:21]: out of curiosity, wouldn't `\`${DEBUG_PREFIX}:verbose\`` work as well?

File: src/webrtc/connection.ts [hunk 4]
```diff
@@ -114,9 +114,33 @@ class WebRTCConnection implements MultiaddrConnection {
 
     this.source = getAbortableSource(this.createSource(), this.options?.signal)
 
-    this.sink = this._sink.bind(this)
+    let sinkCreator: Promise<void>
```
Comment (tolbrino) [inline src/webrtc/connection.ts:117]: Please document the entire block of logic.

File: src/webrtc/connection.ts [hunk 5]
```diff
@@ -206,104 +221,106 @@ class WebRTCConnection implements MultiaddrConnection {
...
-          async function* (this: WebRTCConnection): Stream['source'] {
-            let webRTCFinished = false
-
-            let result: SinkType
-
-            const next = () => {
-              sourcePromise = (source as Stream['source']).next()
-              result = undefined
-            }
+    await new Promise<void>((resolve, reject) =>
```
Comment (tolbrino) [inline src/webrtc/connection.ts:224]: Please document this block of logic more. The way its implemented is fairly complex.

File: src/relay/context.ts [hunk 6]
```diff
@@ -60,14 +60,31 @@ class RelayContext extends EventEmitter {
 
     this.source = this.createSource()
 
+    let sinkCreator: Promise<void>
```
Comment (tolbrino) [inline src/relay/context.ts:63]: Please add documentation to this block of logic.

File: src/relay/connection.ts [hunk 7]
```diff
@@ -159,9 +159,35 @@ class RelayConnection extends EventEmitter implements MultiaddrConnection {
 
     this.source = this.createSource()
 
-    this._stream.sink(this.sinkFunction())
+    let sinkCreator: Promise<void>
```
Comment (tolbrino) [inline src/relay/connection.ts:162]: Please add documentation to this block of logic.

**Review by tolbrino — APPROVED** [review]
I leave it up to you if you document in a separate PR. But it really needs more documentation.

File: src/webrtc/connection.ts (hunk 3 above)
Comment (robertkiel) [reply src/webrtc/connection.ts:21]: think so, also easier to read. changing it

File: src/relay/connection.ts (hunk 2 above)
Comment (robertkiel) [reply src/relay/connection.ts]: changed

File: src/relay/connection.spec.ts (hunk 1 above)
Comment (robertkiel) [reply src/relay/connection.spec.ts]: changed

File: src/webrtc/connection.ts (hunk 5 above)
Comment (robertkiel) [reply src/webrtc/connection.ts:225]: added a few comments

File: src/relay/context.ts (hunk 6 above)
Comment (robertkiel) [reply src/relay/context.ts:66]: done

File: src/relay/connection.ts (hunk 7 above)
Comment (robertkiel) [reply src/relay/connection.ts:166]: done

**Comment by robertkiel** [comment]
Documentation will be improved once changing to proper FSM

**Pull request merged**